File: hotel-front/hotel_pets/lib/controller/tutor_controller.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import './services/api.dart';

class TutorController {
  final Api api = Api();
  List<Map<String, dynamic>> tutores = [];

  Future<void> createTutor(Map<String, dynamic> tutor) async {
    try {
      final response = await api.postRequest("tutor", tutor);

      print("chegou a resposta");
      if (response.statusCode == 403) {
        throw Exception("Insira um email válido");
      }
      if (response.statusCode != 201) {
        throw Exception("Erro ao criar o tutor");
      }
    } catch (e) {
      if (e is http.ClientException) {
        throw Exception('Não foi possível conectar');
      } else {
        rethrow;
      }
    }
  }

  Future<void> getTutores() async {
    try {
      final response = await api.getRequest("tutor");

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        tutores = data.cast<Map<String, dynamic>>();
      } else {
        throw Exception('Erro ao buscar tutores');
      }
    } catch (e) {
      if (e is http.ClientException) {
        throw Exception('Não foi possível conectar');
      } else {
        rethrow;
      }
    }
  }

  Future<void> updateTutor(int id, Map<String, dynamic> tutor) async {
    try {
      final response = await api.putRequest("tutor/${id}", tutor);
      if (response.statusCode != 200 && response.statusCode != 201) {
        print('erro ao salvar o tutor');
        throw Error();
      }
    } catch (e) {
      if (e is http.ClientException) {
        throw Exception('Não foi possível conectar');
      } else {
        rethrow;
      }
    }
  }

  Future<void> deleteTutor(int id) async {
    try {
      final response = await api.deleteRequest("tutor/${id}");
      if (response.statusCode == 409) {
        print("asfasfasdkfadsfjasjfnadsjfhasjdf");
        throw Exception(
            "Não foi possível apagar o tutor. Tente apagar todos os animais dependentes.");
      }
      if (response.statusCode != 200) {
        print('erro ao deletar');
        throw Error();
      }
    } catch (e) {
      if (e is http.ClientException) {
        throw Exception('Não foi possível conectar');
      } else {
        rethrow;
      }
    }
  }
}

File: hotel-back/src/controller/tutorController.js
import tutorService from "../services/tutorService.js";

const listarTutores = async (request, response) => {
    try {
        const tutor = await tutorService.listAllTutores();
        response.json(tutor);
    } catch(error) {
        response.status(500).json({error: error.message});
    }
};

const criarTutor = async (request, response) => {
    try {
        console.log(request.body);
        const {nome, email}= request.body;
        console.log(nome + '   --   ' + email + '   ---  teste')
        const tutor = await tutorService.createTutor(nome, email);
        response.status(201).json(tutor);
    } catch(error) {
        response.status(400).json({error: error.message});
    }
};

const atualizarTutor = async (request, response) => {
    try {
        const { id } = request.params;
        const dadosAtualizados = request.body;

        const res = await tutorService.updateTutor(id, dadosAtualizados);
        response.status(200).json(res);
    } catch(err) {
        response.status(400).json({error: err.message});
    }
};

const apagarTutor = async (request, response) => {
    try {
        const {id} = request.params;
        const res = await tutorService.deleteTutor(id);

        if(res.message == "SequelizeForeignKeyConstraintError") {
            return response.status(409).json({message: "Não foi possível apagar o tutor. Tente apagar todos os animais dependentes."});
        }
        response.status(200).json(res);
    } catch(err) {
        response.status(400).json({message: err.message});
    }
}

export default {listarTutores, criarTutor, atualizarTutor, apagarTutor};